compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUninitializedEnumChecker.kt

```diff
@@ -0,0 +1,166 @@
+        val enumEntries = enumClass.declarations.filterIsInstance<FirEnumEntry>()
+
+        // Members inside the companion object of an enum class
+        if (calleeContainingClass == enumClass.companionObject) {
+            // Uninitialized from the point of view of members or enum entries of that enum class
+            if (accessedContext in enumMemberProperties || accessedContext in enumEntries) {
+                if (calleeDeclaration is FirProperty) {
+                    // From KT-11769
+                    // enum class Fruit(...) {
+                    //   APPLE(...);
+                    //   companion object {
+                    //     val common = ...
+                    //   }
+                    //   val score = ... <!>common<!>
+                    // }
+                    reporter.reportOn(source, FirErrors.UNINITIALIZED_VARIABLE, calleeDeclaration.symbol, context)
+                } else {
+                    // enum class EnumCompanion2(...) {
+                    //   INSTANCE(<!>foo<!>())
+                    //   companion object {
+                    //     fun foo = ...
+                    //   }
+                    // }
+                    // <!>Companion<!>.foo() v.s. <!>foo<!>()
+                    if ((expression.explicitReceiver as? FirResolvedQualifier)?.symbol?.fir == enumClass.companionObject) {
+                        reporter.reportOn(
+                            expression.explicitReceiver!!.source,
+                            FirErrors.UNINITIALIZED_ENUM_COMPANION,
+                            enumClass.symbol,
+                            context
+                        )
+                    } else {
+                        reporter.reportOn(source, FirErrors.UNINITIALIZED_ENUM_COMPANION, enumClass.symbol, context)
+                    }
+                }
+            }
+        }
+
+        // The enum entries of an enum class
+        if (calleeDeclaration in enumEntries) {
+            val calleeEnumEntry = calleeDeclaration as FirEnumEntry
+            // Uninitialized from the point of view of members of that enum class
+            if (accessedContext in enumMemberProperties) {
+                // From KT-6054
+                // enum class MyEnum {
+                //   A, B;
+                //   val x = when(this) {
+                //     <!>A<!> -> ...
+                //     <!>B<!> -> ...
+                //   }
+                // }
+                reporter.reportOn(source, FirErrors.UNINITIALIZED_ENUM_ENTRY, calleeEnumEntry.symbol, context)
+            }
+            // enum class D(...) {
+            //   D1(<!>D2<!>),
+            //   D2(...)
+            // }
+            if (accessedContext in enumEntries && context.containingDeclarations.lastOrNull()?.isEnumEntryInitializer == true) {
+                // Technically, this is equal to `enumEntries.indexOf(accessedContext) < enumEntries.indexOf(calleeDeclaration)`.
```

Comment:
I believe this should be `<=` (in the index comparison)

Reply:
Ah, self reference, thanks!